### ISSUE TITLE
Does $nodesOfType access a stale state?

### ISSUE THREAD

**Issue by zurfyx** [facebook/lexical#4509]
TL;DR: working example with the weird behaviour: https://codesandbox.io/s/stale-nodes-of-type-lwdg3z?file=/src/UpdateTrackerPlugin.tsx

---

In my React app at work I have a plugin that register an update listener and extract all nodes of a certain type to process them outside the editor. Thing is, I noticed that $nodesOfType looks like it's accessing some stale state, while $dfs always reads from the latest available state.

As a minimal reproduction example, in the linked CodeSandbox I have a sample plugin (the UpdateTrackerPlugin) that listens for state updates and collects the CustomNode nodes using both $nodesOfType and $dfs.

Collected nodes texts are then shown in the sections below the text editor.

Is it something I'm not handling on my own? Or a concept I'm missing? Or it's simply some unintended behaviour?

---

Source: morgandam @ https://discord.com/channels/953974421008293909/1103258229988929627
